golang/cosmos/ante/inbound.go

```diff
@@ -10,83 +10,116 @@ import (
-			// mempool. To mitigate this, Swingset should compute the number of
-			// messages allowed for mempool admission as if its max queue length
-			// was lower (e.g. 50%). This is the QueueInboundMempool entry.
-			if inboundsAllowed == -1 {
-				state := ia.sk.GetState(ctx)
-				entry := swingtypes.QueueInbound
-				if ctx.IsCheckTx() {
-					entry = swingtypes.QueueInboundMempool
-				}
-				allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, entry)
-				if found {
-					actions, err := ia.sk.ActionQueueLength(ctx)
-					if err != nil {
-						return ctx, err
-					}
-					if actions < allowed {
-						inboundsAllowed = allowed - actions
-					} else {
-						inboundsAllowed = 0
-					}
-				} else {
-					// if number of allowed entries not given, fail closed
-					inboundsAllowed = 0
-				}
-			}
-			if inboundsAllowed > 0 {
-				inboundsAllowed--
-			} else {
-				defer func() {
-					telemetry.IncrCounterWithLabels(
-						[]string{"tx", "ante", "inbound_not_allowed"},
-						1,
-						[]metrics.Label{
-							telemetry.NewLabel("msg", sdk.MsgTypeURL(msg)),
-						},
-					)
-				}()
-				return ctx, ErrInboundQueueFull
+		inbounds := inboundMessages(msg)
+		if inbounds == 0 {
+			continue
+		}
+		if inboundsAllowed == -1 {
+			var err error
+			inboundsAllowed, err = ia.allowedInbound(ctx)
+			if err != nil {
+				return ctx, err
 			}
 		}
+		if inboundsAllowed >= inbounds {
+			inboundsAllowed -= inbounds
+		} else {
+			defer func() {
+				telemetry.IncrCounterWithLabels(
+					[]string{"tx", "ante", "inbound_not_allowed"},
+					1,
+					[]metrics.Label{
+						telemetry.NewLabel("msg", sdk.MsgTypeURL(msg)),
+					},
+				)
+			}()
+			return ctx, ErrInboundQueueFull
+		}
 	}
 	return next(ctx, tx, simulate)
 }
+
+// allowedInbound returns the allowed number of inbound queue messages or an error.
+// Look up the limit from the swingset state queue sizes: from QueueInboundMempool
+// if we're running CheckTx (for the hysteresis described above), otherwise QueueAllowed.
+func (ia inboundAnte) allowedInbound(ctx sdk.Context) (int32, error) {
+	state := ia.sk.GetState(ctx)
+	entry := swingtypes.QueueInbound
+	if ctx.IsCheckTx() {
+		entry = swingtypes.QueueInboundMempool
+	}
+	allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, entry)
+	if !found {
+		// if number of allowed entries not given, fail closed
+		return 0, nil
+	}
+	actions, err := ia.sk.ActionQueueLength(ctx)
+	if err != nil {
+		return 0, err
+	}
+	if actions >= allowed {
+		return 0, nil
+	}
+	allowed -= actions
+	if allowed > maxInboundPerTx {
+		return maxInboundPerTx, nil
+	}
+	return allowed, nil
+}
+
+// inboundMessages returns the nunber of inbound queue messages in msg.
+func inboundMessages(msg sdk.Msg) int32 {
+	switch msg.(type) {
+	case *swingtypes.MsgDeliverInbound,
+		*swingtypes.MsgInstallBundle,
+		*swingtypes.MsgProvision,
+		*swingtypes.MsgWalletAction,
+		*swingtypes.MsgWalletSpendAction:
+		return 1
+	}
+	return 0
+}
```

Comment:
Please put this functionality somewhere in `x/swingset/types`.

I'd prefer using an interface that describes `GetInboundMsgCount() int32` which the appropriate messages could implement.  That would encourage people to copy-pasta when creating new transaction messages, and thereby not forget to update this list.

Reply:
We should probably quickly sync about the right interface, since I assume we'll need to extend this or write a new ante handler to do the priority checks and `ctx` decoration needed for #6576.  At first it'll be a simple type check, but we'll also need more logic in the future where we check if the sender is allowed to send messages of that particular type (likely against a list stored in vstorage).

Reply:
Oh, we already have the `ControllerAdmissionMsg` interface with a `CheckAdmissibility` method.  It would be trivial to add a second method there to count the number of queue entries used by the message type.

Reply:
BTW, `CheckAdmissibility` is the right place for additional logic, since it has access to the `x/swingset` Keeper.  We can modify the `ControllerAdmissionMsg` interface as much as we'd like, it's a feature that's not exposed outside the Agoric Golang code.

The only reason I proposed adding another method is that the code in this PR is doing something slightly different from the `CheckAdmissibility` path I was carving out.

Reply:
`MsgProvision` is intentionally omitted from `vm.ControllerAdmissionMsg`, but needs to implement `GetInboundMsgCount()`. How would you like to handle that? I could create a new interface `InboundMsgCarrier` containing `GetInboundMsgCount()`, and have `ControllerAdmissionMsg` inherit that.

Reply:
Did as proposed above. PTAL.